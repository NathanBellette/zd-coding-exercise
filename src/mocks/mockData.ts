import {Plan, Product, Subscription} from '../interfaces';

export const plans: Plan[] = [
    {
        id: '1',
        name: 'Basic',
        cost: 1
    },
    {
        id: '2',
        name: 'Good',
        cost: 10
    },
    {
        id: '3',
        name: 'Better',
        cost: 100
    },
    {
        id: '4',
        name: 'Best',
        cost: 1000
    }
];

const products: Product[] = [
    {
        id: '1',
        plans: plans
    }
]

export const subscriptions: Subscription[] = [
    {
        id: '1',
        product: products[0],
        plan: plans[0],
        seats: 5,
        cost: 5
    }
];