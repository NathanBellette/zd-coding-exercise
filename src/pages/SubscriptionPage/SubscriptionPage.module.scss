.container {
  display: flex;
  flex-direction: row;
  justify-content: center;

  .subscription {
    display: flex;
    flex-direction: column;
    width: 50%;

    header {
      height: 5rem;
      margin-top: 1rem;
    }

    article {
      background-color: white;
      padding: 15px;

      .inputRow {
        display: flex;
        height: 12rem;
        width: 100%;
        margin-top: 3rem;
        justify-content: space-evenly;

        @media only screen and (max-width: 600px) {
          flex-direction: column;
        }

        .planSelect {
          display: block;
          width: 100px;
        }
      }
    }

    .buttonRow {
      display: flex;
      flex-direction: row;
      justify-content: flex-end;
      margin-top: 1rem;
    }
  }
}