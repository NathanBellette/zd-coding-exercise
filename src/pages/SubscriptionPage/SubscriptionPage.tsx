import React, {useEffect, useState} from 'react';
import Select from 'react-select';
import {useToasts} from 'react-toast-notifications';
import {Plan, Subscription} from '../../common/interfaces';
import TextField from '../../components/TextField/TextField';
import TextFieldDisplay from '../../components/TextFieldDisplay/TextFieldDisplay';
import Button from '../../components/Button/Button';
import styles from './SubscriptionPage.module.scss';
import axios from 'axios';
import {useHistory, useParams} from 'react-router-dom';
import {shouldDisableUpdate} from '../../common/helpers';
import Loading from "../../components/Loading/Loading";

export interface ComponentProps {
    currentSubscription: Subscription | undefined;
    previewSubscription: Subscription | undefined;
    setPreviewSubscription: (subscription: Subscription) => void;
    setCurrentSubscription: (subscription: Subscription) => void;
}

const SubscriptionPage: React.FC<ComponentProps> = ({
                                                        currentSubscription,
                                                        previewSubscription,
                                                        setPreviewSubscription,
                                                        setCurrentSubscription}) => {
    type Params = {
      id: string;
    };
    const {id} = useParams<Params>();
    const history = useHistory();
    const [loading, setLoading] = useState<boolean>(false);
    const {addToast}  = useToasts();

    useEffect(() => {
        setLoading(true)
        axios.get('/api/current/1')
            .then(response => {
                setLoading(false);
                setCurrentSubscription(response.data.subscription);
                setPreviewSubscription(response.data.subscription);
            })
            .catch(error => {
                setLoading(false);
                addToast(error.message, {
                    appearance: 'error',
                    autoDismiss: false
                });
            })
            .finally(() => setLoading(false));
    }, []);

    const getSubscriptionPreview = (previewSubscription: Subscription) => {
        setLoading(true);
        axios.put('/api/preview/1', previewSubscription)
            .then(response => {
                setLoading(false);
                setPreviewSubscription(response.data);
            })
    };

    const handlePlanChange = (option: Plan | null) => {
        if(option == null) return;
        const selectedPlan = currentSubscription?.product.plans.find(p => p.id === option.id);

        if(previewSubscription && selectedPlan){
            getSubscriptionPreview({
                ...previewSubscription,
                plan: selectedPlan
            })
        }
    };

    const handleSeatsChange = (e: React.FormEvent<HTMLInputElement>) => {
        const numSeats = e.currentTarget.value;
        if(previewSubscription) {
            getSubscriptionPreview({
                ...previewSubscription,
                seats: Number(numSeats)
            });
        }
    };

    const handleUpdateClicked = () => {
        if(previewSubscription) {
            axios.post('/api/current/1', previewSubscription).then(response => {
                setPreviewSubscription(response.data.subscription);
            });
            history.push('/confirmation');
        }
    };

    const currencyFornatter = new Intl.NumberFormat('en-AU', {
        style: 'currency',
        currency: 'AUD',
        maximumFractionDigits: 0,
        minimumFractionDigits: 0
    });

    return (
        <div className={styles.container}>
            <section className={styles.subscription}>
                <header>
                    <h2>Subscription</h2>
                </header>
                <article>
                    <Loading loading={loading} />
                    <div className={styles.inputRow}>
                        <div className={styles.planSelect}>
                            <Select
                                options={previewSubscription?.product.plans}
                                value={previewSubscription?.plan}
                                getOptionLabel={option => option.name}
                                getOptionValue={option => option.id}
                                onChange={handlePlanChange} />
                        </div>
                        <TextField id="numSeats" name="numSeats" label="Seats" value={previewSubscription?.seats?.toString()} onChange={handleSeatsChange} />
                        <TextFieldDisplay label="Price" value={currencyFornatter.format(previewSubscription?.cost || 0)} />
                    </div>
                </article>
                <div className={styles.buttonRow}>
                    <Button onClick={handleUpdateClicked} disabled={shouldDisableUpdate(previewSubscription, currentSubscription)}>Update Subscription</Button>
                </div>
            </section>
        </div>
    )
}

export default SubscriptionPage;
