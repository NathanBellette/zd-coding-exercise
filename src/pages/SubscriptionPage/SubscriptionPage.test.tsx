import React from 'react';
import {render} from '@testing-library/react';
import SubscriptionPage, {ComponentProps} from './SubscriptionPage';
import {subscriptions} from '../../mocks/mockData';
import {ToastProvider} from "react-toast-notifications";

describe('<SubscriptionPage />', () => {
   const defaultProps: ComponentProps = {
       currentSubscription: subscriptions[0],
       previewSubscription: { ...subscriptions[0] },
       setPreviewSubscription: jest.fn(),
       setCurrentSubscription: jest.fn()
   };

   it('renders the component', () => {
       const { container } = render(<ToastProvider><SubscriptionPage {...defaultProps} /></ToastProvider>);
      expect(container).toMatchSnapshot();
   });
})