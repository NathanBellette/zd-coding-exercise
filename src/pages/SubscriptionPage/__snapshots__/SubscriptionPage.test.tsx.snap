// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubscriptionPage /> renders the component 1`] = `
<div>
  <div
    class="loading"
  >
    <span
      class="css-1xdhyk6"
    >
      <span
        class="css-1puqkdb"
      />
      <span
        class="css-10y34i5"
      />
      <span
        class="css-1puqkdb"
      />
    </span>
  </div>
</div>
`;
